src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,453 @@
+    if (!timeDiff)
+        timeDiff = 1;
+
+    // this is the distance doable by the player in 1 sec, using the time done to move to this point.
+    uint32 clientSpeedRate = distance2D * 1000 / timeDiff;
+
+    // we did the (uint32) cast to accept a margin of tolerance
+    if (clientSpeedRate > speedRate)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Speed-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+bool FlyHackCheck::OnCheck(Player* player, MovementInfo const& /* movementInfo */, uint16 /*opcode = 0*/) const
+{
+    if (!player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_FLYING))
+        return false;
+
+    if (player->HasAuraType(SPELL_AURA_FLY) ||
+        player->HasAuraType(SPELL_AURA_MOD_INCREASE_MOUNTED_FLIGHT_SPEED) ||
+        player->HasAuraType(SPELL_AURA_MOD_INCREASE_FLIGHT_SPEED))
+        return false;
+
+    TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Fly-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+    return true;
+}
+
+bool WalkOnWaterCheck::OnCheck(Player* player, MovementInfo const& /* movementInfo */, uint16 /*opcode = 0*/) const
+{
+    if (!player->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_WATERWALKING))
+        return false;
+
+    // if we are a ghost we can walk on water
+    if (!player->IsAlive())
+        return false;
+
+    if (player->HasAuraType(SPELL_AURA_FEATHER_FALL) ||
+        player->HasAuraType(SPELL_AURA_SAFE_FALL) ||
+        player->HasAuraType(SPELL_AURA_WATER_WALK))
+        return false;
+
+    TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Walk on Water - Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+    return true;
+}
+
+bool JumpHackCheck::OnCheck(Player* player, MovementInfo const& /*movementInfo*/, uint16 opcode /*= 0*/) const
+{
+    AnticheatData* playerData = GetPlayerData(player->GetGUID());
+    if (!playerData)
+        return false;
+
+    if (playerData->GetLastOpcode() == MSG_MOVE_JUMP && opcode == MSG_MOVE_JUMP)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Jump-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+bool TeleportPlaneCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    if (player->m_movementInfo.pos.GetPositionZ() != 0 ||
+        movementInfo.pos.GetPositionZ() != 0)
+        return false;
+
+    if (movementInfo.HasMovementFlag(MOVEMENTFLAG_FALLING))
+        return false;
+
+    //DEAD_FALLING was deprecated
+    //if (player->getDeathState() == DEAD_FALLING)
+    //    return;
+    float x, y, z;
+    player->GetPosition(x, y, z);
+    float ground_Z = player->GetMap()->GetHeight(x, y, z);
```

Comment:
Well, the actual "hack checking" code is the only thing remnant from the original; any suggestions here, besides throttle? Maybe caching the value elsewhere (is it good idea?)